packages/perf/src/index.ts

```diff
@@ -0,0 +1,172 @@
+  writeBlockSize: BigInt(64 << 10)
+}
+
+export interface PerfService {
+  perf: (connection: Connection, sendBytes: bigint, recvBytes: bigint, options?: AbortOptions) => Promise<void>
+  measureDownloadBandwidth: (connection: Connection, size: bigint) => Promise<number>
+  measureUploadBandwidth: (connection: Connection, size: bigint) => Promise<number>
+}
+
+export interface PerfServiceInit {
+  protocolName?: string
+  maxInboundStreams?: number
+  maxOutboundStreams?: number
+  timeout?: number
+  writeBlockSize?: bigint
+}
+
+export interface PerfServiceComponents {
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DefaultPerfService implements Startable, PerfService {
+  public readonly protocol: string
+  private readonly components: PerfServiceComponents
+  private started: boolean
+  private readonly databuf: ArrayBuffer
+  private readonly maxInboundStreams: number
+  private readonly maxOutboundStreams: number
+  private readonly timeout: number
+  private readonly writeBlockSize: bigint
+
+  constructor (components: PerfServiceComponents, init: PerfServiceInit) {
+    this.components = components
+    this.started = false
+    this.protocol = init.protocolName ?? PROTOCOL_NAME
+    this.databuf = new ArrayBuffer(Number(init.writeBlockSize ?? WRITE_BLOCK_SIZE))
+    this.maxInboundStreams = init.maxInboundStreams ?? MAX_INBOUND_STREAMS
+    this.maxOutboundStreams = init.maxOutboundStreams ?? MAX_INBOUND_STREAMS
+    this.timeout = init.timeout ?? TIMEOUT
+    this.writeBlockSize = init.writeBlockSize ?? WRITE_BLOCK_SIZE
+  }
+
+  async start (): Promise<void> {
+    await this.components.registrar.handle(this.protocol, (data: IncomingStreamData) => { void this.handleMessage(data) }, {
+      maxInboundStreams: this.maxInboundStreams,
+      maxOutboundStreams: this.maxOutboundStreams
+    })
+    this.started = true
+  }
+
+  async stop (): Promise<void> {
+    await this.components.registrar.unhandle(this.protocol)
+    this.started = false
+  }
+
+  isStarted (): boolean {
+    return this.started
+  }
+
+  async handleMessage (data: IncomingStreamData): Promise<void> {
+    const { stream } = data
+
+    const writeBlockSize = this.writeBlockSize
+
+    let bytesToSendBack: bigint | null = null
+
+    for await (const buf of stream.source) {
+      if (bytesToSendBack === null) {
+        bytesToSendBack = BigInt(buf.getBigUint64(0, false))
+      }
+      // Ingest all the bufs and wait for the read side to close
+    }
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    if (bytesToSendBack === null) {
+      throw new Error('bytesToSendBack was null')
+    }
+
+    await stream.sink(async function * () {
+      while (bytesToSendBack > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > bytesToSendBack) {
+          toSend = bytesToSendBack
+        }
+        bytesToSendBack = bytesToSendBack - toSend
+        yield uint8Buf.subarray(0, Number(toSend))
```

Comment:
Both random bytes and zeros is fine by the specification. See past discussion by you, @achingbrain.

https://github.com/libp2p/specs/pull/478#discussion_r1037393837

All other implementations use zeroes.

Reply:
Zero is fine, random bytes can also be fine (as long as it's not a different set of random bytes, because then you may be limited by how fast you can generate random bytes).

Reply:
Look at me making the same point twice 😆 

> you may be limited by how fast you can generate random bytes

You can generate them ahead of time so this is solvable.  I haven't measured it but in my head at least any network operation will be orders of magnitude slower than a pRNG?

Reply:
Let's measure it! On my machine (M1 Max):

```
$  dd if=/dev/urandom of=/dev/null bs=1500 count=100
100+0 records in
100+0 records out
150000 bytes (150 kB, 146 KiB) copied, 0.000466 s, 322 MB/s
```

That's 2.5 Gigabits/s, while fast is not fast enough to saturate a fairly common 10 gbps line.

(Yes you can fiddle with blocksizes, but I think 1500 is fair as it's a common MTU size.)

But what's the issue with sending zero bytes? Is it that you're afraid we'll silently introduce compression on a transport and thus not correctly measure throughput? If that's the case sending zero bytes is actually ideal because it would be obvious if there is compression happening.

There's also the question of arbitrarily compressing at a transport level. That kind of thing opens you up the [TLS CRIME](https://en.wikipedia.org/wiki/CRIME) style attacks. So I don't think we should support compression without being application aware.